### PR TITLE
feat: Implement permission trie for RBAC authorization optimization

### LINKED ISSUE #58: RBAC Authorization Performance Optimization
## Problem Statement

The current RBAC authorization system in kube-oidc-proxy suffers from significant performance issues due to its nested loop structure. This becomes a bottleneck in large Kubernetes clusters with thousands of RBAC rules, causing latency spikes and poor scalability.

## Current Implementation Issues

The existing authorization flow involves:
1. Visiting rules for each binding 
2. Processing bindings through nested loops
3. Checking if binding applies
4. Getting role rules 
5. Checking each rule 

This results in poor performance that degrades as the RBAC configuration grows.

## Proposed Solution: Permission Trie Structure

Implement a hierarchical permission trie that mirrors the Kubernetes resource model:

```
User → Cluster → Namespace → Api Group → Resource → Verb
```

### Authorization Flow

```mermaid
graph TD
    A[Authorization Request] --> B[Find User Node]
    B --> C[Find Cluster Node<br/>check wildcard *]
    C --> D[Find Namespace Node<br/>check wildcard *]
    D --> E[Find Resource Node<br/>check wildcard *]
    E --> F[Check Verb<br/>check wildcard *]
    F -- Verb exists --> G[Allow]
    F -- Verb not found --> H[Deny]
```

### Elimination of Nested Loops

The trie structure eliminates the need for nested loops by providing:
- **O(1)** lookup for user permissions
- **Constant time** authorization regardless of RBAC rule count
- **Hierarchical wildcard support** (* for clusters, namespaces, resources, verbs)

## Implementation Requirements

1. **New PermissionTrie package** with hierarchical structure
2. **RBACAuthorizer interface** implementing the new authorization logic
3. **Integration with existing cluster manager**
4. **Comprehensive test coverage** including edge cases and wildcards
5. **Backward compatibility** with existing RBAC configurations

### Key Features
- **Thread-safe** implementation with read-write mutex
- **Memory efficient** node cleanup when permissions are removed
- **Wildcard support** at all levels (cluster, namespace, apiGroup, resource, verb)
- **Resource name specificity** support for fine-grained permissions
- **Non-resource URL** permission checking

## Expected Impact

- **90-99% reduction** in authorization latency
- **Constant time performance** regardless of cluster size
- **Improved scalability** for large Kubernetes deployments
- **Reduced CPU/memory usage** for authorization operations

This optimization will significantly improve the performance and scalability of kube-oidc-proxy in enterprise environments with complex RBAC configurations.

### PR DESCRIPTION
**Pull request by husenkuresh1**
This PR implements a permission trie structure to replace the nested loop-based RBAC authorization system, reducing authorization time.

## Fixes Issue

Closes #58 

## Changes proposed
- Created permissiontrie.go with new hierarchical permission structure
- Updated authorizer.go with the new RBAC implementation
- Enhanced clustermanager.go for permission checking integration
- Removed legacy authorizer from cluster.go
- Added comprehensive tests in authorizer_test.go and permissiontrie_test.go

## Backward Compatibility

- Maintains existing API interfaces
- Supports all existing RBAC configurations  
- No breaking changes to external APIs
- Preserves all functionality while improving performance

## Check List (Check all the applicable boxes) 

- [ ] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] This PR does not contain plagiarized content.
- [ ] The title of my pull request is a short description of the requested changes.

### CODE REVIEW

File: pkg/util/authorizer/permissiontrie.go [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+	"post":             VerbPost,
+	"put":              VerbPut,
+	"*":                VerbAll,
+}
+
+// PermissionTrie holds all permissions in a tree-like structure for fast lookups.
+type PermissionTrie struct {
+	subjectNodes map[string]*SubjectNode
+	mu           sync.RWMutex
+}
```
Comment (deltasquare4) [inline pkg/util/authorizer/permissiontrie.go:46]: I do not understand how this is a Trie. Can you please explain what happens here? Thanks.
